phpseclib/Net/SFTP.php

```diff
@@ -961,6 +972,108 @@ private function readlist($dir, $raw = true)
+        switch ($this->packet_type) {
+            case NET_SFTP_HANDLE:
+                // http://tools.ietf.org/html/draft-ietf-secsh-filexfer-13#section-9.2
+                // since 'handle' is the last field in the SSH_FXP_HANDLE packet, we'll just remove the first four bytes that
+                // represent the length of the string and leave it at that
+                $handle = substr($response, 4);
+                break;
+            case NET_SFTP_STATUS:
+                // presumably SSH_FX_NO_SUCH_FILE or SSH_FX_PERMISSION_DENIED
+                $this->logError($response);
+                return false;
+            default:
+                throw new \UnexpectedValueException('Expected NET_SFTP_HANDLE or NET_SFTP_STATUS. '
+                    . 'Got packet type: ' . $this->packet_type);
+        }
+
+        $this->update_stat_cache($dir, []);
+
+        try {
+            while (true) {
+                // http://tools.ietf.org/html/draft-ietf-secsh-filexfer-13#section-8.2.2
+                // why multiple SSH_FXP_READDIR packets would be sent when the response to a single one can span arbitrarily many
+                // SSH_MSG_CHANNEL_DATA messages is not known to me.
+                if (!$this->send_sftp_packet(NET_SFTP_READDIR, Strings::packSSH2('s', $handle))) {
+                    return false;
+                }
+
+                $response = $this->get_sftp_packet();
+                switch ($this->packet_type) {
+                    case NET_SFTP_NAME:
+                        list($count) = Strings::unpackSSH2('N', $response);
+                        for ($i = 0; $i < $count; $i++) {
+                            list($shortname, $longname) = Strings::unpackSSH2('ss', $response);
+                            $attributes = $this->parseAttributes($response);
+                            if (!isset($attributes['type'])) {
+                                $fileType = $this->parseLongname($longname);
+                                if ($fileType) {
+                                    $attributes['type'] = $fileType;
+                                }
+                            }
+                            $content = $attributes + ['filename' => $shortname];
+
+                            if (isset($attributes['type']) && $attributes['type'] == NET_SFTP_TYPE_DIRECTORY && ($shortname != '.' && $shortname != '..')) {
+                                $this->update_stat_cache($dir . '/' . $shortname, []);
+                            } else {
+                                if ($shortname == '..') {
+                                    $temp = $this->realpath($dir . '/..') . '/.';
+                                } else {
+                                    $temp = $dir . '/' . $shortname;
+                                }
+                                $this->update_stat_cache($temp, (object) ['lstat' => $attributes]);
+                            }
+                            yield $content;
```

Comment:
Instead of filling the array yield the value for processing and pause the function until next() is called.

Reply:
I think this function needs to be more DRY, but that would probably require some refactoring of existing code. My trail of thought was to introduce a new feature, but maybe a generator for these kind of requests should be default for all functions. Implementation thereof could be of any kind. Filling an array with the attributes, which costs allot of memory would merely be a helper, but not advised, since the default would be a generator/Iterable.